### PR TITLE
Initial crossdock behavior list

### PR DESCRIPTION
**Pull request by breerly**
Enumerating our initial cross-language test suite. This should verify the core promises of YARPC, before we move on to higher level features.

@yarpc/yarpc

### CODE REVIEW

File: crossdock.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+Crossdock Tests
```
Comment (kriskowal) [inline crossdock.md]: Perhaps merely "YARPC Test Harness Behaviors". Can and should be exercised with and without Crossdock.

File: crossdock.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+----
+
+For each transport, make a call to the json `echo` procedure, returning a list of results.
+
+```
+200 OK
+
+[
+    {"status": "passed", "output": "made tchannel json request to echo - got resp: 34bdchk"},
+    {"status": "passed", "output": "made http json request to echo - got resp: 34bdchk"},
```
Comment (kriskowal) [inline crossdock.md]: I suspect we will want to capture the relevant cross section of the parameters on these response objects so we can filter and build tables various ways.

**Comment by kriskowal** [comment]
Comment lost in rebase: I suspect we will want to capture the relevant cross section of the parameters on these response objects so we can filter and build tables various ways.

File: crossdock.md (hunk 1 above)
Comment (breerly) [inline crossdock.md]: These test cases should run as part of the regular test suite - so the distinction shouldn't matter much at this point.

Comment (breerly) [inline crossdock.md]: In this case, I'm specifically spelling out what the responses should look like, so I think this title fits.

Comment (kriskowal) [inline crossdock.md]: :+1:

File: crossdock.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
+Crossdock Tests
+===============
+
+Participating YARPC libraries must ship a Crossdock client with the following behaviors implemented.
+
+
+raw
+---
+
+For each transport, make a call to the `echo` procedure, returning a list of results.
```
Comment (kriskowal) [inline crossdock.md]: Unless we do a refactor, this is currently `echo/raw`.

File: crossdock.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    {"status": "passed", "output": "made http raw request to echo - got resp: 34bdchk"},
+]
+```
+
+Dimensions: `server`, `transport`
+
+json
+----
+
+For each transport, make a call to the `echo-token` json procedure, returning a list of results.
```
Comment (kriskowal) [inline crossdock.md]: This is currently merely `echo`.

File: crossdock.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+]
+```
+
+Dimensions: `server`, `transport`
+
+thrift
+------
+
+For each transport, make a call to every Thrift procedure in [ThriftTest.thrift](ThriftTest.thrift),
+returning a list of results for each call.
```
Comment (kriskowal) [inline crossdock.md]: I recommend that we do something less exhaustive for the first pass, e.g., the Echo::echo test I’m proposing for the Thrift MVP. Maybe rename this `thrift-coverage`.

File: crossdock.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+application-headers
+-------------------
+
+For each transport and each encoding, make a call to verify:
+
+1. Sending valid headesr
+2. Sending non-string values
+3. Sending empty string values
+1. Sending no headers, eg an empty map
+2. Varying key casing - to ensure case insensitivity
```
Comment (kriskowal) [inline crossdock.md]: We’ll eventually need to also do an `context-headers` pass. I think it’s safe to assume this is a subset.

File: crossdock.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    {"status": "passed", "output": "made a tchannel+json call to verify valid headers, got {'this': 'worked'}"},
+    {"status": "passed", "output": "made a tchannel+thrift call to verify valid headers, got {'this': 'worked'}"},
+    {"status": "passed", "output": "made a http+raw call to verify valid headers, got {'this': 'worked'}"},
+    {"status": "passed", "output": "made a http+json call to verify valid headers, got {'this': 'worked'}"},
+    {"status": "passed", "output": "made a http+thrift call to verify valid headers, got {'this': 'worked'}"},
+]
+
+```
+
+Dimensions: `server`, `transport`, `encoding`
```
Comment (kriskowal) [inline crossdock.md]: Will need to specifically hit `echo/raw`, `echo`, and `Echo::echo` to exercise these, based on the encoding. A note to implementers to make a map from encoding to some reusable "echo" caller code.

**Comment by kriskowal** [comment]
:ok: land and iterate

File: crossdock.md (hunk 4 above)
Comment (breerly) [inline crossdock.md]: Was hoping we could rename it... although that might be a PITA.

Comment (breerly) [inline crossdock.md]: Actually, as long as we each switch our sever to be `echo/raw` then we can move the json impl to `echo-token` right...

File: crossdock.md (hunk 6 above)
Comment (breerly) [inline crossdock.md]: Perhaps - once we get this foundational stuff tested, we can move onto context.

File: crossdock.md (hunk 4 above)
Comment (kriskowal) [inline crossdock.md]: I’m not sure the alternative names are more clear than the existing names. Not that I’m satisfied, but I don’t expect to regret this name scheme. Name collisions between raw and JSON are rare, but I expect JSON to win a namespace fight and for raw to be special. I don’t think -token distinguishes JSON better than /raw distinguishes raw.

**Pull request merged**